contrib/storage-kafka/src/main/java/org/apache/drill/exec/store/kafka/KafkaGroupScan.java

```diff
@@ -262,6 +269,17 @@ public void applyAssignments(List<DrillbitEndpoint> incomingEndpoints) {
     assignments = AssignmentCreator.getMappings(incomingEndpoints, Lists.newArrayList(partitionWorkMap.values()));
   }
 
+  @Override
+  public GroupScan applyLimit(int maxRecords) {
+    records = maxRecords; // Just apply the limit value into sub-scan
+    return super.applyLimit(maxRecords);
+  }
+
```

Comment:
I'm fairly certain this isn't actually working.  The super method always returns `null`. 

https://github.com/apache/drill/blob/53e6f2697b31cb76a38e75fe95283512715d28c3/exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/AbstractGroupScan.java#L174-L178

What I think you need to do here is:
1.  Create a clone constructor in the `KafkaGroupScan`
2. Call that and return a new `KafkaGroupScan` with the limit applied.

Reply:
I wouldn't recommend that.  My understanding of how all this works is that Calcite/Drill create multiple copies of the GroupScan to try to find the optimal query plan.   The GroupScans have to be immutable otherwise strange things will happen when you execute more complex queries, and in all likelihood, the limit won't be pushed down.

Reply:
Follow your suggestion, create the immutable groupscan always.